README.md
![linux](https://github.com/BlondeBurrito/plotrs/actions/workflows/build_linux.yml/badge.svg)
![windows](https://github.com/BlondeBurrito/plotrs/actions/workflows/build_windows.yml/badge.svg)
[![crates.io](https://img.shields.io/crates/v/plotrs.svg)](https://crates.io/crates/plotrs)
[![docs](https://img.shields.io/badge/docs-docs.rs-orange.svg)](https://docs.rs/plotrs)
[![MIT/Apache 2.0](https://img.shields.io/badge/license-MIT%2FApache-blue.svg)](./LICENSE)

# plotrs

A CLI app for plotting csv data sets onto a graph. It works by reading a graph definition from a `.ron` file, then extracts data from one or more csv files and produces a `.png` image. Currently only scatter graphs are supported.

<img src="examples/scatter_full/showcasing_a_bunch_of_cool_features.png" alt="s" width="800"/>

Back in the mists of time I used to use [GNU Octave](https://www.gnu.org/software/octave/index) for plotting data about plasmonic absorption and photovoltaic-thermoelectric currents. As part of my Rust journey I thought I'd try writting a program for plotting data points in a similar style.

## Features

* Overlay best fit curves onto your graph
* Graph element/component positions and sizes are dynamically calculated based on the size of the image you want
* Multiple colours and symbols can be used to plot data sets
* Data can be sourced from one or more csv files - you're simply targeting certain columns in a given file for extraction
* Error bars - plot uncertainty in `x` and `y` singly or jointly
* Appropriate quadrants are drawn if your data makes use of negative `x-y` values

## Install

`cargo install plotrs`

## How To Use

Create a `.ron` file containing the configuration of your desired chart and generate a `png` with:

```bash
plotrs -g <graph_type> -c <path_to_config_ron_file> -o <dir_for_output_png>
```

E.g

```bash
plotrs -g scatter -c scatter_config.ron -o here/please
```

Note that if your canvas is too small then your title and axis labels may become blurry.

## Graph `.ron` Schemas

### Scatter Definition

```rust
Scatter(
	title: "Engery against Time for Fuzzing About Things",
	canvas_pixel_size: (840, 600),
	x_axis_label: "Time (ms)",
	x_axis_resolution: 11, // Number of times the x-axis will be divided to show your data scale
	y_axis_label: "Energy (kJ)",
	y_axis_resolution: 11, // Number of times the y-axis will be divided to show your data scale
	has_grid: false, // Should the graph have a light grey background grid
	has_legend: false, // should a legend be generated? Only really useful with multiple data sets
	// data sets can be sourced from the same csv or from different ones and each can be configured with different colours/symbols
	data_sets: [
		DataSet(
			data_path: "scatter.csv",
			has_headers: true, // if your data has headers set to `true` so they can be ignored
			x_axis_csv_column: 0, // which column contains the x values
			x_axis_error_bar_csv_column: None, // which column contains x uncertainty Some(usize) or None
			y_axis_csv_column: 1, // which column contains the y values
			y_axis_error_bar_csv_column: None, // which column contains y uncertainty Some(usize) or None
			name: "Very interesting", // legend will indicate which colour and symbol correspond to which data set
			colour: Orange, // the colour to render a data point
			symbol: Cross, // the shape a plotted data point should take
			symbol_radius: 5, // The size of a drawn symbol in (1+ symbol_radius) pixels
			symbol_thickness: 0, // The thinkness of a drawn symbol in (1 + symbol_thickness) pixels
			best_fit: None, // A curve to fit to the axes. Some(BestFit) or None
		),
	],
)
```

Where your `csv` data may look like (note the lack of whitespace between columns!):

```csv
x,y
0.5,0.5
1.0,1.0
1.5,1.5
```

In a directory you may have:

```txt
- my_config.ron
- data.csv
```

So to generate a `png` you'd run from within the directory `plotrs -g scatter -c my_config.ron` and it'll write a `png` next to the files.

## Symbol Types/Colours

The following symbols can be used for plotting data points:

* Cross
* Circle
* Triangle
* Square
* Point

With the following colours:

* White
* Black
* Grey
* Orange
* Red
* Blue
* Green
* Pink

## [Best Fit Schemas](https://github.com/BlondeBurrito/plotrs/tree/main/examples/scatter_best_fit)

Each data set definition can also specify a Best Fit line to be drawn. In the examples below the data sets are tiny and the symbols are coloured white to hide them in the background canvas, they really just define the extent of the axes to show case overlaying a Best Fit.

### Linear

```rust
y = gradient * x + y_intercept
```

`Some(Linear(gradient: 1.0, y_intercept: 0.0, colour: Black))`

<img src="examples/scatter_best_fit/linear_best_fit.png" alt="s" width="210"/>

### Quadratic

```rust
y = intercept + (linear_coeff * x) + (quadratic_coeff * x.powf(2))
```

`Some(Quadratic(intercept: 1.0, linear_coeff: 0.0, quadratic_coeff: 1.0, colour: Black))`

<img src="examples/scatter_best_fit/quadratic_best_fit.png" alt="s" width="210"/>

### Cubic

```rust
y = intercept + (linear_coeff * x) + (quadratic_coeff * x.powf(2)) + + (cubic_coeff * x.powf(3))
```

`Some(Cubic(intercept: 1.0, linear_coeff: -0.5, quadratic_coeff: 1.0, cubic_coeff: 1.0, colour: Black))`

<img src="examples/scatter_best_fit/cubic_best_fit.png" alt="s" width="210"/>

### Generic Polynomial

For custom polynomials you supply a map of coefficients where each key is the `nth` power `x` will be raised by and the value is the coefficient it'll be multiplied by.

Roughly:

```rust
for (k, v) in coefficients.iter() {
	y += v * x.powf(k);
}
```

The following extends the Cubic best fit into a Quartic Polynomial:

`Some(GenericPolynomial(coefficients: {0: 1.0, 1: 1.0, 2: 1.0, 3: 1.0, 4: -1.0}, colour: Black))`

Which to the human eye kinda looks like: `1 + x + x^2 + x^3 - x^4`.

<img src="examples/scatter_best_fit/generic_polynomial_best_fit.png" alt="s" width="210"/>

### Exponential

```rust
y = (constant * base.powf(power * x)) + vertical_shift;
```

`Some(Exponential(constant: 0.5, base: 2.7, power: -1.0, vertical_shift: 3.0, colour: Black))`

<img src="examples/scatter_best_fit/exponential_best_fit.png" alt="s" width="210"/>

### Gaussian

```rust
`y = (variance * (2.0 * PI).sqrt()).powf(-1.0) * E.powf(-(x - expected_value).powf(2.0) / (2.0 * variance.powf(2.0)))`
```

`Some(Gaussian(expected_value: 0.0, variance: 0.3, colour: Black))`

<img src="examples/scatter_best_fit/gaussian_best_fit.png" alt="s" width="210"/>

### Sinusoidal

```rust
y = amplitude * ((period * x) + phase_shift).sin() + vertical_shift;
```

`Some(Sine(amplitude: 2.0, period: 1.0, phase_shift: 0.0, vertical_shift: 3.0, colour: Black))`

<img src="examples/scatter_best_fit/sinusoidal_best_fit.png" alt="s" width="210"/>

### Cosinusoidal

```rust
y = amplitude * ((period * x) + phase_shift).cos() + vertical_shift;
```

`Some(Cosine(amplitude: 2.0, period: 1.0, phase_shift: 0.0, vertical_shift: 3.0, colour: Black))`

<img src="examples/scatter_best_fit/cosinusoidal_best_fit.png" alt="s" width="210"/>

## Examples

### [Simple Scatter](https://github.com/BlondeBurrito/plotrs/tree/main/examples/scatter)

<img src="examples/scatter/engery_against_time_for_fuzzing_about_things.png" alt="s" width="400"/>

### [Image Size Scales Elements Dynamically](https://github.com/BlondeBurrito/plotrs/tree/main/examples/scatter_large)

Based on the dimensions of your image (`canvas_size`) the text and axes positions are automatically calculated. You can also toggle a light grey background grid drawn the from axes scales.

<img src="examples/scatter_large/a_large_graph_for_testing_positions_of_graph_elements_and_stuff.png" alt="s" width="800"/>

### [Scatter Multidata](https://github.com/BlondeBurrito/plotrs/tree/main/examples/scatter_multidata_one_csv)

From single or multiple `csv` files you can plot several data sets onto a single graph. Each data set can be configured to plot with a different colour and/or symbol. The legend can be toggled on and off. The size and thickness of the symbols are configurable on a per data set basis.

From a single `csv` containing multiple columns for different data sets:

<img src="examples/scatter_multidata_one_csv/oh_wow__multiple_data_sets.png" alt="s" width="800"/>

From two `csv` files where each contains a column pair:

<img src="examples/scatter_multidata_two_csv/data_from_two_csv_files_woweeeee.png" alt="s" width="800"/>

### [Scatter Error Bars](https://github.com/BlondeBurrito/plotrs/tree/main/examples/scatter_error_bars)

You can also indicate uncertainty with the use of error bars which can be specified for either axes.

<img src="examples/scatter_error_bars/we_have_some_certainty_in_y.png" alt="s" width="800"/>
<img src="examples/scatter_error_bars/we_have_some_certainty_in_x_and_y.png" alt="s" width="800"/>

### [Quadrants Derived From Data](https://github.com/BlondeBurrito/plotrs/tree/main/examples/scatter_quadrants)

Based on the range of values across a given number of data sets the cartesian quadrants required are determined during exection with scale markings and axis labels moved appropriately.

<img src="examples/scatter_quadrants/top_right_quadrant_example.png" alt="s" width="300"/><img src="examples/scatter_quadrants/top_left_quadrant_example.png" alt="s" width="300"/><img src="examples/scatter_quadrants/bottom_right_quadrant_example.png" alt="s" width="300"/><img src="examples/scatter_quadrants/bottom_left_quadrant_example.png" alt="s" width="300"/><img src="examples/scatter_quadrants/top_pair_quadrant_example.png" alt="s" width="300"/><img src="examples/scatter_quadrants/bottom_pair_quadrant_example.png" alt="s" width="300"/><img src="examples/scatter_quadrants/left_pair_quadrant_example.png" alt="s" width="300"/><img src="examples/scatter_quadrants/right_pair_quadrant_example.png" alt="s" width="300"/><img src="examples/scatter_quadrants/all_quadrants_example.png" alt="s" width="300"/>

## Troubleshooting

### The numbers along the axis are long floats overlapping one another

Try changing the `x` and `y` axis resolutions to numbers which are a factor of your largest values + 10%. What happens under the hood is that the largest values in your data set are found and slightly scaled so that data points avoid being plotted directly on an axis and thus obscurring some text/markers. When an axis is drawn it has a certain length in pixels and the resolution decides how many times it gets chopped up to display scale markers. To map a data value (f32) to a pixel (u32) there is a conversion where a single pixel represents some amount or length of value data. For an awkward resolution the pixel length between two scale markers could be a long float rather than rounded whole number.

E.g if the largest `x` value in your data is `10` try setting the `x_axis_resolution` to `10 * 1.1 = 11`, that should produce `11` nice scale markers with whole numbers. Likewise a resolution `22` would produce nice markers also as `11` fits into `22` snugly.

### The title/axis labels/legend are blurry

Try increasing the size of your canvas if the edges of the text become blurry.

## Contributing

* If you're unsure about something raise an issue first
* Fork it
* Tippy tap your keyboard
* Submit a PR

## LICENSE

[Dual license of MIT and Apache](https://github.com/BlondeBurrito/plotrs/blob/main/LICENSE).

## TODO

* Show BestFit types in legend
* Allow overriding font
* checked sub and addition to ensure pixel u32s are not overflowing maybe?
* Split/simplify drawing methods out and then add a billion tests, many around position calculations
* What methods/modules can be reused to draw other graph types...


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file with the power of `git-cliff` changelog generation.

## [0.1.3] - 2022-04-28

### Bug Fixes

- Git cliff tag matching now locates existing tags

## [0.1.2] - 2022-04-28

### Bug Fixes

- Set cargo metadata categories

## [0.1.0] - 2022-04-28

### Bug Fixes

- Fixed axis scales to accurately mark data points
- Legend element now laign correctly
- More reliably calculate the pixels making up the circle symbol
- Recalculate legend position and how it occupies space on the canvas
- X-axis scale markers now sit outside the axis in whitespace when negative y range is dominant
- Reposition y axis scale markers to avoid graph overlap and make use of free whitespace
- Axis scale text no longer overlays where negative and positive axes meet

### Documentation

- Added module documentation where it was lacking
- Main crate docs

### Features

- Big bang
- Added pipelines
- Added second example to test large canvas sizes with graph positioning
- New ron scatter data structure
- Font size is dynamically calculated based on canvas size
- Read and parse csv from DataSets into data points
- Split canvas.rs into multiple sub-modules
- Plots cross symbols onto canvas
- Ron config sets axis resolution for drawing data labels along the scale
- Ron config sets axis resolution for drawing data labels along the scale
- Positions of all graph elements are trul dynamic and determined from canvas size
- Added legend
- Added error bar plotting
- Added multidata example extracting sets from different files
- Added square symbol
- Added triangle symbol
- Added mini-markers to sit between axis scales. The number of mini-markers is auto generated based on the amount of pixel space available between the main scale markers
- Added linear best fit plotting
- Added quadratic, cubic, exponential, sinusoidal and cosinusoial best fit curves
- Added Generic Polynomial best fit line
- Added example for each best fit curve type
- Prototype generating relevant quadrants to build based on x-y ranges in csv data
- All quadrants drawn in various combinations
- Added gaussian distribution best fit type
- Support for negative axes

### Miscellaneous Tasks

- Workflow runs examples with trace logging

### Refactor

- Split axes functions into modules for easier readability
- Reduced number of match arms for finding label positions as some quadrants make use of the same positions
- Make clippy happy

<!-- generated by git-cliff -->


src/scatter/mod.rs
//! Constructs a scatter graph based on the `.ron` definition:
//!
//! ```txt
//! Scatter(
//!    title: "Engery against Time for Fuzzing About Things",
//!    canvas_pixel_size: (840, 600),
//!    x_axis_label: "Time (ms)",
//!    x_axis_resolution: 11, // Number of times the x-axis will be divided to show your data scale
//!    y_axis_label: "Energy (kJ)",
//!    y_axis_resolution: 11, // Number of times the y-axis will be divided to show your data scale
//!    has_grid: false, // Should the graph have a light grey background grid
//!    has_legend: false, // should a legend be generated? Only really useful with multiple data sets
//!    // data sets can be sourced from the same csv or from different ones and each can be configured with different colours/symbols
//!    data_sets: [
//!        DataSet(
//!            data_path: "scatter.csv",
//!            has_headers: true, // if your data has headers set to `true` so they can be ignored
//!            x_axis_csv_column: 0, // which column contains the x values
//!            x_axis_error_bar_csv_column: None, // which column contains x uncertainty Some(usize) or None
//!            y_axis_csv_column: 1, // which column contains the y values
//!            y_axis_error_bar_csv_column: None, // which column contains y uncertainty Some(usize) or None
//!            name: "Very interesting", // legend will indicate which colour and symbol correspond to which data set
//!            colour: Orange, // the colour to render a data point
//!            symbol: Cross, // the shape a plotted data point should take
//!            symbol_radius: 5, // The size of a drawn symbol in (1+ symbol_radius) pixels
//!            symbol_thickness: 0, // The thinkness of a drawn symbol in (1 + symbol_thickness) pixels
//!            best_fit: None, // A curve to fit to the axes. Some(BestFit) or None
//!        ),
//!    ],
//! )
//! ```

use ron::de::from_reader;
use serde::Deserialize;
use std::fs::File;
use tracing::{debug, error, info};

mod data;

use crate::{
	canvas::{
		axes::axis_x::build_x_axis_label,
		axes::axis_y::build_y_axis_label,
		axes::{
			axis_x::get_x_axis_pixel_length, axis_y::get_y_axis_pixel_length, draw_xy_axes,
			get_xy_axis_pixel_min_max, get_xy_axis_pixel_origin,
		},
		best_fit::BestFit,
		draw_base_canvas,
		glyphs::FontSizes,
		legend::build_legend,
		plot::DataSymbol,
		quadrants::get_quadrants,
		save_image,
		title::build_title,
		VHConsumedCanvasSpace,
	},
	colours::*,
	scatter::data::{build_data_points, get_data_bounds, get_legend_fields},
};
/// Specification of a scatter graph
#[derive(Debug, Deserialize)]
struct Scatter {
	/// The title of the graph
	title: String,
	/// Image size in pixels
	canvas_pixel_size: (u32, u32),
	/// X-axis label
	x_axis_label: String,
	/// Number of times the x-axis will be divided to show your data scale. Advised to make it a ratio of your largest x value
	x_axis_resolution: u32,
	/// Y-axis label
	y_axis_label: String,
	/// Number of times the y-axis will be divided to show your data scale. Advised to make it a ratio of your largest y value
	y_axis_resolution: u32,
	/// Should the graph has a light grey background grid
	has_grid: bool,
	/// Should a legend be generated
	has_legend: bool,
	/// Defines where the data is and which parts to use
	data_sets: Vec<DataSet>,
}
/// The source of each data set and how it should be represented
#[derive(Debug, Deserialize)]
pub struct DataSet {
	/// Path to csv data
	data_path: String,
	/// Does the csv contain headers
	has_headers: bool,
	/// Which column in the csv contains the x-axis data
	x_axis_csv_column: usize,
	/// Optional, the column which contains an uncertainty measure
	x_axis_error_bar_csv_column: Option<usize>,
	/// Which column in the csv contains the y-axis data
	y_axis_csv_column: usize,
	/// Optional, the csv column which contains an uncertainty measure
	y_axis_error_bar_csv_column: Option<usize>,
	/// Name of the data set, useful when generating a legend to distinguish sets
	name: String,
	/// The colour a data point should be plotted as
	colour: Colour,
	/// The shape used to represent the data point
	symbol: DataSymbol,
	/// The size of a drawn symbol in (1+ symbol_radius) pixels
	symbol_radius: u32,
	/// The thinkness of a drawn symbol in (1 + symbol_thickness) pixels
	symbol_thickness: u32,
	/// Optional, a type of best fit line to draw
	best_fit: Option<BestFit>,
}

/// Creates a canvas and draws the scatter graph over it
pub fn scatter_builder(path: &str, output: &str, csv_delimiter: &str) {
	info!("Building scatter chart...");
	let scatter: Scatter = Scatter::deserialise(path);
	info!("Drawing canvas...");
	let mut canvas = draw_base_canvas(scatter.canvas_pixel_size);
	// Calcualte font sizes
	info!("Calculating font sizes...");
	let font_sizes = FontSizes::new(&scatter.canvas_pixel_size);
	// To fit the various labels, axes, legend and title all onto the canvas snugly we need some values
	// telling us how much space has already been occupied by previous elements.
	// We use these 4 values of the struct to indicate the amount of pixel space consumed from each border of the canvas
	// in vertical and horizontal directions. The general convention is the are arranged clockwise from the top
	let mut canvas_edges_used = VHConsumedCanvasSpace::new();
	// Place the title at the top of the canvas and tell us how many v-pixels have been used
	info!("Building title...");
	// Building the title tells us how many pixel have been consumed from the top of the canvas
	canvas_edges_used.add(build_title(
		&mut canvas,
		&scatter.title,
		font_sizes.title_font_size,
	));
	// optionally build the legend
	if scatter.has_legend {
		let legend_fields = get_legend_fields(&scatter.data_sets);
		let legend_origin_x = canvas.dimensions().0
			- canvas_edges_used.h_space_from_right
			- (canvas.dimensions().0 / 10);
		let legend_origin_y = canvas_edges_used.v_space_from_top + canvas.dimensions().1 / 4;
		// let legend_origin: (u32, u32) = (axis_max.0, axis_max.1 * 2);
		canvas_edges_used.add(build_legend(
			&mut canvas,
			(legend_origin_x, legend_origin_y),
			legend_fields,
			font_sizes.legend_font_size,
		));
	}
	// Find the size of the data - this tells us whether any axis requires a negative range.
	// Of the form `(min_x, min_y), (max_x, max_y)`
	info!("Finding min and max range of data...");
	let (min_xy, max_xy): ((f32, f32), (f32, f32)) =
		get_data_bounds(&scatter.data_sets, csv_delimiter);
	// We want to create buffer space around our bounds so data points are not plotted directly on an axis, if
	// large symbols are used for plotting they may obscure data labels on an axis. We scale the bounds by 10%
	// taking care to shrink and expand bounds based on their sign
	let min_x_scaled = if min_xy.0.is_sign_positive() {
		min_xy.0 / 1.1
	} else {
		min_xy.0 * 1.1
	};
	let min_y_scaled = if min_xy.1.is_sign_positive() {
		min_xy.1 / 1.1
	} else {
		min_xy.1 * 1.1
	};
	let max_x_scaled = if max_xy.0.is_sign_positive() {
		max_xy.0 * 1.1
	} else {
		max_xy.0 / 1.1
	};
	let max_y_scaled = if max_xy.1.is_sign_positive() {
		max_xy.1 * 1.1
	} else {
		max_xy.1 / 1.1
	};
	let min_xy_scaled = (min_x_scaled as i32, min_y_scaled as i32);
	debug!("Minimum x-y with buffer space {:?}", min_xy_scaled);
	let max_xy_scaled = (max_x_scaled as i32, max_y_scaled as i32);
	debug!("Maximum x-y with buffer space {:?}", max_xy_scaled);
	// With the scaled values we find which quadrants to draw are valid
	let quadrants = get_quadrants(min_xy_scaled, max_xy_scaled);
	info!("Quadrants to draw based on data set {:?}", quadrants);
	info!("Building y-axis label...");
	// Draws the y-axis label and returns the amount of pixel space used up by the glyphs
	canvas_edges_used.add(build_y_axis_label(
		&mut canvas,
		scatter.y_axis_label,
		font_sizes.axis_font_size,
		&quadrants,
		canvas_edges_used.v_space_from_top,
		canvas_edges_used.h_space_from_right,
		canvas_edges_used.v_space_from_bottom,
		canvas_edges_used.h_space_from_left,
	));
	info!("Building x-axis label...");
	// Draws the x-axis label and returns the amount of pixel ocupied from the bottom
	canvas_edges_used.add(build_x_axis_label(
		&mut canvas,
		scatter.x_axis_label,
		font_sizes.axis_font_size,
		&quadrants,
		canvas_edges_used.v_space_from_top,
		canvas_edges_used.h_space_from_right,
		canvas_edges_used.v_space_from_bottom,
		canvas_edges_used.h_space_from_left,
	));
	// With the text drawn we can calculate the rectangular space for the axes, represrnted as two tuples
	// pinpointing the bottom left origin of the graph and the top right corner.
	// Pixel position showing the maximum extents of the axes
	let (axis_min, axis_max): ((u32, u32), (u32, u32)) = get_xy_axis_pixel_min_max(
		&quadrants,
		canvas_edges_used.v_space_from_top,
		canvas_edges_used.h_space_from_right,
		canvas_edges_used.v_space_from_bottom,
		canvas_edges_used.h_space_from_left,
		canvas.dimensions(),
		scatter.x_axis_resolution,
		scatter.y_axis_resolution,
	);
	debug!("Minimum axis placement {:?}", axis_min);
	debug!("Maximun axis placement {:?}", axis_max);
	// Pixel position of axes origin can be determined from the min-max intersection
	let axis_origin: (u32, u32) = get_xy_axis_pixel_origin(&quadrants, axis_min, axis_max);
	debug!("Origin axis placement {:?}", axis_origin);
	// We need to know how the csv data scales to the length of axes for plotting,
	// ie. we need a scale factor of how many units of data there is to one pixel
	// First we need the axis length
	let x_axis_length = get_x_axis_pixel_length(axis_min.0, axis_max.0);
	// Y-axis max is in fact a smaller number due to canvas image origin
	let y_axis_length = get_y_axis_pixel_length(axis_max.1, axis_min.1);
	debug!("X-axis length {}", x_axis_length);
	debug!("Y-axis length {}", y_axis_length);
	let x_data_min_max_limits: (i32, i32) = (min_xy_scaled.0, max_xy_scaled.0);
	let y_data_min_max_limits: (i32, i32) = (min_xy_scaled.1, max_xy_scaled.1);
	// Now we can find the number of axis units per x and y
	// Ensure we don't divide by zero!
	if !(max_xy_scaled.0 as f32 - min_xy_scaled.0 as f32).is_normal() {
		error!("Difference between the smallest and largest x values have produced Zero, Infinite, NaN or a Subnormal value. Likely if your data set only contains a single row. Ensure you have multiple rows and that your largest x value minus your smallest x doesn't produce zero");
		std::process::exit(1)
	}
	let x_axis_data_scale_factor: f32 =
		x_axis_length as f32 / (max_xy_scaled.0 as f32 - min_xy_scaled.0 as f32).abs();
	if !(max_xy_scaled.1 as f32 - min_xy_scaled.1 as f32).is_normal() {
		error!("Difference between the smallest and largest y values have produced Zero, Infinite, NaN or a Subnormal value. Likely if your data set only contains a single row. Ensure you have multiple rows and that your largest y value minus your smallest y doesn't produce zero");
		std::process::exit(1)
	}
	let y_axis_data_scale_factor: f32 =
		y_axis_length as f32 / (max_xy_scaled.1 as f32 - min_xy_scaled.1 as f32).abs();
	debug!("X-axis scale factor {}", x_axis_data_scale_factor);
	debug!("Y-axis scale factor {}", y_axis_data_scale_factor);

	draw_xy_axes(
		&quadrants,
		&mut canvas,
		axis_origin,
		axis_min,
		axis_max,
		x_axis_length,
		y_axis_length,
		x_data_min_max_limits,
		y_data_min_max_limits,
		font_sizes.axis_unit_font_size,
		scatter.has_grid,
		scatter.x_axis_resolution,
		scatter.y_axis_resolution,
	);
	// if a line of best fit has been specified then draw it
	for set in &scatter.data_sets {
		match &set.best_fit {
			Some(curve) => {
				info!("Plotting best fit...");
				let points = curve.find_coordinates(
					x_data_min_max_limits.0,
					x_data_min_max_limits.1,
					y_data_min_max_limits.0,
					y_data_min_max_limits.1,
					scatter.canvas_pixel_size.0 as i32 * 2,
				);
				let origin_offset = (axis_origin.0, axis_origin.1);
				for p in points.iter() {
					p.draw_point(
						&mut canvas,
						x_axis_data_scale_factor,
						y_axis_data_scale_factor,
						origin_offset,
					);
				}
			}
			None => {}
		}
	}
	// get the csv data content and plot it
	build_data_points(
		&scatter.data_sets,
		csv_delimiter,
		&mut canvas,
		x_axis_data_scale_factor,
		y_axis_data_scale_factor,
		(axis_origin.0, axis_origin.1),
	);

	// save the resulting image
	save_image(canvas, output, scatter.title);
}

impl Scatter {
	/// Based on a path deserialise a `.ron` file into a graph data structure
	fn deserialise(path: &str) -> Scatter {
		// attempt to open the .ron file
		let f = match File::open(path) {
			Ok(file) => file,
			Err(e) => {
				error!("Failed to open .ron file at {}, error: {:?}", path, e);
				std::process::exit(1)
			}
		};
		// attempt to deserialise the config data
		let scatter: Scatter = match from_reader(f) {
			Ok(x) => x,
			Err(e) => {
				error!(
					"Failed to load config, maybe you're missing a comma? Error: {}",
					e
				);
				std::process::exit(1);
			}
		};
		debug!("Ron config {:?}", &scatter);
		scatter
	}
}
